### DISCUSSION TITLE
How to use lv_observer in lvgl 9.x

### DISCUSSION THREAD

**Discussion by sukesh-ak** [sukesh-ak/BSP-IDF5-ESP_LCD-LVGL9#4] [Show and tell]
Here is an example of how to use `lv_observer` in `LVGL 9`
https://github.com/sukesh-ak/BSP-IDF5-ESP_LCD-LVGL9/blob/main/main/main_ui.c

It's used for the brightness percentage slider to show the percentage value in label as well as use it to change the actual brightness of the display using `callback` from `lv_subject_add_observer_obj`.

### Callback function

```c
// callback function
static void brightness_observer_cb(lv_observer_t * observer, lv_subject_t * subject)
{
    int32_t brightness_percent = lv_subject_get_int(subject);
    bsp_display_brightness_set(brightness_percent);
}
```

### UI Binding using lv_observer
```c
lv_subject_t brightness_subject;

// Initialize lv_subject into int (int32_t) type
lv_subject_init_int(&brightness_subject, 80);

/*Create a slider in the center of the display*/
lv_obj_t * slider = lv_slider_create(lv_screen_active());
lv_obj_align(slider,LV_ALIGN_CENTER,0,10);
lv_slider_bind_value(slider, &brightness_subject);                  // Bind slider value with the lv_subject
lv_obj_set_style_anim_duration(slider, 2000, 0);

/*Create a label below the slider*/
lv_obj_t * slider_label = lv_label_create(lv_screen_active());
lv_obj_align_to(slider_label, slider, LV_ALIGN_OUT_BOTTOM_MID, 0, 10);
lv_label_bind_text(slider_label, &brightness_subject, "%d %%");     // Bind label text with lv_subject

// Add lv_observer with a callback when the value changes
lv_subject_add_observer_obj(&brightness_subject, brightness_observer_cb, NULL, NULL);
```

**Comment by sukesh-ak** [comment]
@mhaberler I thought you might be interested in this.

**Comment by mhaberler** [comment]
thanks - tracking ;)

**Comment by sukesh-ak** [comment]
@kisvegabor ESP32-S3 based sample project showcasing `lv_observer` from `lvgl 9.x`

**Reply by kisvegabor** [reply]
Cool! Happy to see the observer in action! :blush:

**Reply by mhaberler** [reply]
@kisvegabor  - while I have your ear:

I'd be interested on your take on the idea pushing JSON to observers 

like described here: https://github.com/sukesh-ak/BSP-IDF5-ESP_LCD-LVGL9/discussions/6#discussioncomment-9008345

- pretty easy to dump stringified JSON from sources like MQTT, TCP, websockets, Serial ff
- makes a nice interface cut in an application if several different UI's are an option (like additional  a web UI or somesuch)
- not the fastest kid on the block due to serialize/deserialize on every event
- the the other direction (originating from UI) is not as obvious - could be obserervers or queues

I'v also tried pushing ArduinoJson objects directly but Benoit will likely have a heart attack if he sees this
I'm open to other JSON libraries, is there a specific preference in LVGL land?

If you bless the idea (or a variation thereof), happy to do an example for reference

what do you think about an additional observer type for this case?

Michael

**Reply by kisvegabor** [reply]
Hi,
Could you show an example for an observer JSON file?

**Reply by mhaberler** [reply]
let me separate what I have in place, and what I think could be done better and/or differently

what is in place and works:

- a single string observer, the sender just writes a string to the observer:  https://github.com/mhaberler/arduino3-playground/blob/lvgl8-observer/src/ui_message.cpp#L19-L27  
- the string is assumed to be legit JSON
- the observer callback parses the string into JSON and does the right thing to UI objects - i.e. contains application logic:  https://github.com/mhaberler/arduino3-playground/blob/lvgl8-observer/src/lv_custom_observers.cpp#L103-L289

- within the same program: on the originating side the message is constructed as an ArduinoJson document, serialized  to a string, and sent to the uiMessage observer:
https://github.com/mhaberler/arduino3-playground/blob/lvgl8-observer/src/nfcreader.cpp#L192-L203 (example for reading an NFC tag)

- if a JSON string comes from "outside" (HTTP/MQTT etc) the string is passeed to the observer as-is

- the message is tagged with a enum message type so the observer knows what type it is: https://github.com/mhaberler/arduino3-playground/blob/lvgl8-observer/src/nfcreader.cpp#L200

- the observer callback inspects the type tag and decides what do with it: https://github.com/mhaberler/arduino3-playground/blob/lvgl8-observer/src/lv_custom_observers.cpp#L119-L138 (again NFC read example)

in summary, it is an abuse of a single string observer for message passing. One could say it's a size 1 message queue without locks.

That said, it works nicely.


the problems with this approach are IMO:
- it is questionable if the observer is the right pattern for what is essentially a message queue without locks.
- there is a potential for multiple concurrent invocations of callbacks if the callback routine is slow - no queuing.
- constant serialisation/deserialisation JSON->string->JSON slows things down a bit
- passing JsonDocument directly is asking for trouble like memory leaks, I tried but I do not think this is a good solution to avoid serialisation overhead: https://github.com/mhaberler/arduino3-playground/blob/lvgl8-observer/src/ui_message.cpp#L9-L17

what can be done better:

assuming the basic idea - passing structured messages through an observer - makes sense, I would do this:

- replace the JSON string by a msgpack wire format buffer - ArduinoJson can serialize to/from msgpack 
- the observer callback decodes the msgpack buffer, the rest is as above
- introduce a new observer type instead of string.

on the sender side, I see three API functions:
1. pass an ArduinoJson document. That gets serialized to a msgpack buffer, then passed to the observer
2. directly pass a msgpack message to the observer
3. function which takes a JSON string as before. This function deserializes JSON, then serializes to msgpack, then pass that to observer.

(1) and (2) could be used within the same program, (3) would be for interacting externally via HTTP/MQTT/Websockets etc.

Advantage of this approach:
- the simple "pass a JSON string to make something happen" option is available, and a faster alternative as well, both could be library code.
- we stay withing what ArduinoJson can do today.
- msgpack has tons of bindings so serialisation/deserialisation is stock code
- msgpack buffers are a better match for object passing than JSON strings.

Disadvantages:
- msgpack will be replaced by CBOR eventually I think, however ArduinoJson doesnt do CBOR (yet?) and the CBOR bindings/library situation looks less appealing

NB even if the consensus were "you better use a proper queue" the ArduinoJson JsonDocument/JSON string/msgpack API functions would still make sense


Michael

**Reply by sukesh-ak** [reply]
I don't think `json serialze/deserialize` should be part of `lv_observer` since it will make the UI freeze and bring more troubles to the normal functioning of `lvgl` on MCU's.

Currently `lv_observer` also avoids the need of `lock/unlock mutex` requirements.

Keep in mind, we need to make sure none of the long processing procedures runs on `UI thread/task` or in a responsive callbacks.

**Reply by kisvegabor** [reply]
@mhaberler Thans for the detailed description. I agree with @sukesh-ak that the JSON part shouldn't be part of LVGL but we need to make it possible to easily integrate the observers into such environment. 

> it is questionable if the observer is the right pattern for what is essentially a message queue without locks.

In [this comment](https://github.com/lvgl/lvgl/issues/5419#issuecomment-2002004526) I have summarized the advantages of the `lv_observer` over `lv_msg` (from LVGL v8). It might be useful when designing this system too.

**Reply by mhaberler** [reply]
you're bang on, I retract my opinion and maintain the opposite position as of now while I recode ;)

it's pushing the envelope using observers as a general IPC mechanism. 

I guess the more appropriate locus of application is within the LVGL task/UI  context proper like @sukesh-ak 's opening example, or the Wifi indicator [here](https://github.com/sukesh-ak/ESP32-LVGL8x-SDSPI/blob/5d631101ce779991ecf40e700ce0d4ecdfc06561/main/main.cpp#L142-L169) 

this also answers the question "do we need observers to handle objects, not just scalars and strings" as: IMO no.

that suggests interaction of the UI task via structured messages like JSON, msgpack ff better happen with separate threadsafe queues or similar - while leaving any heavy lifting like parsing outside the LVGL task/UI space

meaning: this is not an observer-related issue

**Reply by kisvegabor** [reply]
It seems we are on the same page :relaxed:

**Reply by mhaberler** [reply]
I reworked the IPC mechanism into a [ESP-IDF ringbuffer ](https://github.com/mhaberler/arduino3-playground/blob/lvgl8-observer/src/ringbuffer.h)and a [lv_timer callback](https://github.com/mhaberler/arduino3-playground/blob/lvgl8-observer/src/lv_custom_observers.cpp#L304) in the UI context.

works fine, several lvgl locks can be removed because the ringbuffer is threadsafe.

some JSON parsing still present in UI context but looks acceptable to me

**Reply by sukesh-ak** [reply]
Will check it out

**Comment by mhaberler** [comment]
I'm looking into a client which can work with MQTT 

it looks to me the observer pattern is a good match with a MQTT subscription/topic - I'm thinking about how to best match these up in  generic way instead of per-case C++ coding, maybe through a JSON config

btw the broker does not necessarily be somewhere external, see https://github.com/mlesniew/PicoMQTT  (PicoMQTT is currently learning MQTT-over-websockets and that will leave the route open to use a plain website as client as an alternative)

I'd be interested in exchange of ideas/prototypes on the subject 

-m

**Reply by sukesh-ak** [reply]
Not sure if you have seen my write-up on how to configure MQTT here including using `websockets`. 
Client and C++ code snipped also in the repo.
https://github.com/sukesh-ak/setup-mosquitto-with-docker

**Reply by mhaberler** [reply]
using the same combo: my private mosquitto and MQTTX for exploring
MQTTX is great, as is their client example repo

the PicoMQTT use case comes from a different angle: assume there is no net connectivity and you'd want an web app, or some embedded client receive values from an embedded system publishing sensor data

one of the use cases is to have sensorlogger subscribe to a local arduino connected via wifi to the mobile's hotspot, and receive sensor data from there - @tszheichoi is working on MQTT subscribe for that 

@mlesniew

**Reply by sukesh-ak** [reply]
Thanks for the sensor logger pointer.
Building a UI for ESP32-XX display using `lv_observer` is not a big deal. 
I need it in future but currently getting distracted from some pending work.

Running broker is not an issue at all since you can just run it on a `Raspberry Pi` like I plan for my offline scenarios.
For testing, I already have a broker running on my home server.

**Reply by sukesh-ak** [reply]
@mhaberler Btw here is the Espressif ESP-IDF MQTT5 sample client
https://github.com/espressif/esp-idf/tree/master/examples/protocols/mqtt5

**Reply by mhaberler** [reply]
yeah one can always brute-force the broker issue with some local linux vehicle

I wanted to avoid that in the sensorlogger case, so you'd get away with just say an esp32+custom sensors, and sensorlogger - no extra boxes

some sensorlogger deployment locations I'm using are less than ideal for running an extra linux box:

![oe-sox - 1 (4)](https://github.com/sukesh-ak/BSP-IDF5-ESP_LCD-LVGL9/assets/901479/2b19aaf6-9ac5-409f-875c-75caa132da84)

**Reply by sukesh-ak** [reply]
Interesting use case :)

**Reply by sukesh-ak** [reply]
@mhaberler I think the thread topic got hijacked.
Lets open another discussion thread to discuss MQTT.

**Reply by mhaberler** [reply]
ok, admitted - sorry

**Comment by mhaberler** [comment]
re  Espressif ESP-IDF MQTT5 : tried that, works great - for sensorlogger we need some form of broker

however sensorlogger will eventually be able to subscribe to a broker to fetch values and add them to the log (not yet in the current release, but drop Kelvin a note for a TestFlight invitation - should be around shortly)

with that scheme it should be possible to log arbitrary sensors, maybe even ones read and decodeby [Open MQTT Gateway](https://docs.openmqttgateway.com/)

**Reply by mhaberler** [reply]
on using binary JSON representations for queued UI interactions :

it seems the [nlohmann/json](https://github.com/nlohmann/json) library would be the better choice over ArduinoJson - it supports serialisation and deserialisation to/from CBOR, msgpack and a couple of others  out of the box

to make sense, that of course rests on the assumption that decoding and acting on a CBOR or msgpack message in the LVGL context is faster than parsing a string into JSON as I am doing right now 

works nicely on Arduino, [ESP-IDF component here](https://components.espressif.com/components/johboh/espidf-component)

**Reply by sukesh-ak** [reply]
ESP-IDF comes with json (`cjson`), so I used the same for `ESP32-TUX`
https://github.com/espressif/esp-idf/tree/master/components/json
